### PR TITLE
vcpu_hotpluggable: Add vcpu hotpluggable and order tests

### PR DESCRIPTION
**Pull request by Grace-chhu**
<vcpus>
    <vcpu id='0' enabled='yes' hotpluggable='no' order='1'/>
    <vcpu id='1' enabled='yes' hotpluggable='yes' order='3'/>
    <vcpu id='2' enabled='no' hotpluggable='yes'/>
    <vcpu id='3' enabled='yes' hotpluggable='yes' order='2'/>
    <vcpu id='4' enabled='no' hotpluggable='yes'/>
    <vcpu id='5' enabled='yes' hotpluggable='yes' order='4'/>
  </vcpus>

Signed-off-by: Chenli Hu <chhu@redhat.com>

### CODE REVIEW

**Comment by Grace-chhu** [comment]
# avocado run --vt-type libvirt vcpu_hotpluggable
......
TESTS      : 3
 (1/3) type_specific.io-github-autotest-libvirt.vcpu_hotpluggable.possitive_test: PASS (4.18 s)
 (2/3) type_specific.io-github-autotest-libvirt.vcpu_hotpluggable.negative_test.dup_order1: PASS (23.15 s)
 (3/3) type_specific.io-github-autotest-libvirt.vcpu_hotpluggable.negative_test.dup_order2: PASS (21.66 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0

File: libvirt/tests/src/cpu/vcpu_hotpluggable.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+            # Dumpxml
+            dump_xml = virsh.dumpxml(vm_name).stdout.strip()
+
+            # Check libvirtd log
+            for vcpu in vcpu_list:
+                if vcpu['enabled'] == 'yes' and vcpu['hotpluggable'] == 'yes':
+                    cmd = ("cat %s| grep device_add| grep qemuMonitorIOWrite"
+                           "| grep 'vcpu%s'" % (log_path, vcpu['id']))
+                    logging.info("cmd:\n %s" % cmd)
```
Comment (balamuruhans) [inline libvirt/tests/src/cpu/vcpu_hotpluggable.py]: `logging.info("cmd:\n %s", cmd)`

File: libvirt/tests/src/cpu/vcpu_hotpluggable.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        del vmxml.cpuset
+
+        # create vcpu xml
+        vcpu_list = []
+        vcpu = {}
+        en_list = vcpus_enabled.split(",")
+        hotplug_list = vcpus_hotplug.split(",")
+        order_dict = ast.literal_eval(vcpus_order)
+
+        for i in range(vcpus_max):
```
Comment (balamuruhans) [inline libvirt/tests/src/cpu/vcpu_hotpluggable.py]: proper variable name can be used instead of i

**Comment by Grace-chhu** [comment]
I'll add the test for hot-plugging/cold-plugging vcpus in next two weeks.

File: libvirt/tests/src/cpu/vcpu_hotpluggable.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+                # Check libvirtd log
+                for vcpu in vcpu_list:
+                    if vcpu['enabled'] == 'yes' and vcpu['hotpluggable'] == 'yes':
+                        cmd = ("cat %s| grep device_add| grep qemuMonitorIOWrite"
+                               "| grep 'vcpu%s'" % (config_path, vcpu['id']))
+                        ret = process.run(cmd, ignore_status=False, shell=True)
+                        if ret.exit_status != 0:
+                            logging.error("Failed to find lines about enabled vcpu%s"
+                                          "in libvirtd log." % vcpu['id'])
```
Comment (nasastry) [inline libvirt/tests/src/cpu/vcpu_hotpluggable.py]: "in libvirtd log." % vcpu['id'])
should be 
"in libvirtd log.", vcpu['id'])

File: libvirt/tests/src/cpu/vcpu_hotpluggable.py (hunk 2 above)
Comment (Grace-chhu) [inline libvirt/tests/src/cpu/vcpu_hotpluggable.py]: Changed to vcpu_id, thank you!

File: libvirt/tests/src/cpu/vcpu_hotpluggable.py (hunk 3 above)
Comment (Grace-chhu) [inline libvirt/tests/src/cpu/vcpu_hotpluggable.py]: Changed, thank you!

**Comment by balamuruhans** [comment]
@Grace-chhu you can have follow up with https://github.com/autotest/tp-libvirt/issues/998 and @sathnaga working on APIs helps CPU hotplugging/hotunplugging

File: libvirt/tests/src/cpu/vcpu_hotpluggable.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+            vcpu_lines = re.findall(r"VCPU:.*\n", output)
+            logging.info("vcpuinfo vcpu_lines: \n %s", vcpu_lines)
+            if len(vcpu_lines) != vcpus_current:
+                raise exceptions.TestFail("vcpuinfo is not correct.")
+
+            # Check cpu in guest
+            session = vm.wait_for_login(timeout=30)
+            cmd = "lscpu | grep '^CPU(s):'"
+            output = session.cmd_output(cmd).strip()
+            session.close()
```
Comment (sathnaga) [inline libvirt/tests/src/cpu/vcpu_hotpluggable.py]: you can use vm.get_cpu_count() instead

File: libvirt/tests/src/cpu/vcpu_hotpluggable.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+                        if ret.exit_status != 0:
+                            logging.error("Failed to find lines about enabled vcpu%s"
+                                          "in libvirtd log.", vcpu['id'])
+            # Dumpxml
+            dump_xml = virsh.dumpxml(vm_name).stdout.strip()
+            vcpu_items = re.findall(r"vcpu.*", dump_xml)
+
+            # Check guest vcpu count
+            ret = virsh.vcpucount(vm_name, ignore_status=True, debug=True)
+            output = ret.stdout.strip()
```
Comment (sathnaga) [inline libvirt/tests/src/cpu/vcpu_hotpluggable.py]: https://github.com/autotest/tp-libvirt/issues/1002 am working on wirting a library function to get/check the vcpu values in APIs like
vcpucount, vcpuinfo, vcpupin, and other few helper functions that would help us in hotplug and many other usecases, which am planning to post it by next week, you wish to choose if it works out for you, let me know if
you have comments and suggestions to the same. 

E:g:-
```

def get_vcpucount_details(vm, options):
    """
    To get vcpucount output

    :param: vm_name: Name of the vm
    :param: options: options to passed to vcpucount

    :return: dict of vcpucount output values
    """
.....

def check_vcpu_count(vm, exp_vcpu, vcpupin=None, option="", guest_agent=False):
    """
    Check domain vcpu, including vcpucount, vcpuinfo, vcpupin, vcpu number and
    cputune in domain xml, vcpu number inside the domain.

    :param vm: VM object
    :param exp_vcpu: dict of expect vcpu number:
        exp_vcpu['max_config'] = maximum config vcpu nubmer
        exp_vcpu['max_live'] = maximum live vcpu number
        exp_vcpu['cur_config'] = current config vcpu number
        exp_vcpu['cur_live'] = current live vcpu number
        exp_vcpu['guest_live'] = vcpu number inside the domain
    :param vcpupin: A Dict of expect vcpu affinity
    :param option: Option for virsh commands(setvcpu, setvcpus etc)
    :param guest_agent: True if agent present

    :return: True if the exp_vcpu values matches with virsh API values
            False if not
    """
....
....

```

**Comment by will-Do** [comment]
ping @Grace-chhu

**Comment by Grace-chhu** [comment]
@sathnaga, thank you for your review! I'm busy with other projects these days, and will update the script since next Monday......

**Comment by Grace-chhu** [comment]
# avocado run --vt-type libvirt vcpu_hotpluggable
JOB ID     : 60676b884057e146959ad481904a77085357b7fe
JOB LOG    : /root/avocado/job-results/job-2017-08-08T15.21-60676b8/job.log
TESTS      : 7
 (1/7) type_specific.io-github-autotest-libvirt.vcpu_hotpluggable.possitive_test.vcpus_with_order: PASS (26.15 s)
 (2/7) type_specific.io-github-autotest-libvirt.vcpu_hotpluggable.possitive_test.unplug_vcpus.hot_unplug: PASS (31.22 s)
 (3/7) type_specific.io-github-autotest-libvirt.vcpu_hotpluggable.possitive_test.unplug_vcpus.cold_unplug: PASS (31.04 s)
 (4/7) type_specific.io-github-autotest-libvirt.vcpu_hotpluggable.possitive_test.hotplug_vcpus.hot_plug: PASS (30.94 s)
 (5/7) type_specific.io-github-autotest-libvirt.vcpu_hotpluggable.possitive_test.hotplug_vcpus.cold_plug: PASS (30.69 s)
 (6/7) type_specific.io-github-autotest-libvirt.vcpu_hotpluggable.negative_test.dup_order1: PASS (11.03 s)
 (7/7) type_specific.io-github-autotest-libvirt.vcpu_hotpluggable.negative_test.dup_order2: PASS (12.05 s)
RESULTS    : PASS 7 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0
JOB HTML   : /root/avocado/job-results/job-2017-08-08T15.21-60676b8/html/results.html
TESTS TIME : 173.12 s

File: libvirt/tests/src/cpu/vcpu_hotpluggable.py (hunk 5 above)
Comment (kylazhang) [inline libvirt/tests/src/cpu/vcpu_hotpluggable.py]: @Grace-chhu  do you ever try the func that above mentioned?

File: libvirt/tests/src/cpu/vcpu_hotpluggable.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+            # Check cgroup info
+            en_vcpu_list = re.findall(r"vcpu.*enabled=.yes.*", re_dump_xml)
+            for vcpu_sn in range(len(en_vcpu_list)):
+                vcpu_id = en_vcpu_list[vcpu_sn].split("=")[1].split()[0].strip('\'')
+                cmd = ("lscgroup| grep cpuset| grep %s| grep vcpu%s" %
+                       (vm_name[-3:], vcpu_id))
+                ret = process.run(cmd, ignore_status=False, shell=True)
+                if ret.exit_status != 0:
+                    logging.error("Failed to find lines about enabled vcpu%s"
+                                  "in lscgroup info.", vcpu_id)
```
Comment (kylazhang) [inline libvirt/tests/src/cpu/vcpu_hotpluggable.py]: no test.fail or test.error here?

File: libvirt/tests/src/cpu/vcpu_hotpluggable.py [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+            ret = virsh.vcpuinfo(vm_name, ignore_status=True, debug=True)
+            output = ret.stdout.strip()
+            vcpu_lines = re.findall(r"VCPU:.*\n", output)
+            logging.info("vcpuinfo vcpu_lines: \n %s", vcpu_lines)
+            if len(vcpu_lines) != vcpus_current:
+                test.fail("vcpuinfo is not correct.")
+
+            # Check cpu in guest
+            if not utils_misc.check_if_vm_vcpu_match(vcpus_current, vm):
+                test.fail("cpu number in VM is not correct, it should be %s cpus", vcpus_current)
```
Comment (chloerh) [inline libvirt/tests/src/cpu/vcpu_hotpluggable.py]: This line, to format error message, should use '%', not ','

**Comment by waynesun09** [comment]
Pls fix the problems and squash commit

File: libvirt/tests/src/cpu/vcpu_hotpluggable.py (hunk 5 above)
Comment (Grace-chhu) [inline libvirt/tests/src/cpu/vcpu_hotpluggable.py]: Hi, kylazhang

Thank you for review, above PR#1002 is Open, and without commit yet.

Comment (Grace-chhu) [inline libvirt/tests/src/cpu/vcpu_hotpluggable.py]: Hi, kylazhang

Thanks for review, the PR#998 is about the new virsh command setvcpu in rhel7.4.
    setvcpu - attach/detach vcpu or groups of threads
Which is different to the command used in this script: setvcpus
    setvcpus - change number of virtual CPUs

The PR#1002 is "Open", and without commit by far. Thank you!

Regards,
Grace-chhu

File: libvirt/tests/cfg/cpu/vcpu_hotpluggable.cfg [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+            variants:
+                - vcpus_with_order:
+                    config_libvirtd = "yes"
+                - unplug_vcpus:
+                    variants:
+                        - hot_unplug:
+                            set_live_vcpus = "2"
+                        - cold_unplug:
+                            set_config_vcpus = "3"
+                - hotplug_vcpus:
```
Comment (chunfuwen) [inline libvirt/tests/cfg/cpu/vcpu_hotpluggable.cfg]: The title is - hotplug_vcpus,but sub-variants includes cold_plug,hot_plug,so the title is confused with sub variants

File: libvirt/tests/src/cpu/vcpu_hotpluggable.py [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+            # Restart libvirtd to make the changes take effect in libvirt
+            libvirtd.restart()
+
+        # Set vcpu: placement,current,max vcpu
+        vmxml.placement = vcpus_placement
+        vmxml.vcpu = vcpus_max
+        vmxml.current_vcpu = vcpus_current
+        del vmxml.cpuset
+
+        # create vcpu xml with vcpu hotpluggable and order
```
Comment (chunfuwen) [inline libvirt/tests/src/cpu/vcpu_hotpluggable.py]: Capitalize first letter

**Comment by Grace-chhu** [comment]
# avocado run --vt-type libvirt vcpu_hotpluggable
JOB ID     : 0ac25cea210c21b3849b030429f5d1f3cedd0e4c
JOB LOG    : /root/avocado/job-results/job-2017-08-21T17.19-0ac25ce/job.log
TESTS      : 7
 (1/7) type_specific.io-github-autotest-libvirt.vcpu_hotpluggable.possitive_test.vcpus_with_order: PASS (26.45 s)
 (2/7) type_specific.io-github-autotest-libvirt.vcpu_hotpluggable.possitive_test.unplug.live: PASS (32.58 s)
 (3/7) type_specific.io-github-autotest-libvirt.vcpu_hotpluggable.possitive_test.unplug.config: PASS (31.57 s)
 (4/7) type_specific.io-github-autotest-libvirt.vcpu_hotpluggable.possitive_test.plug.live: PASS (31.59 s)
 (5/7) type_specific.io-github-autotest-libvirt.vcpu_hotpluggable.possitive_test.plug.config: PASS (31.10 s)
 (6/7) type_specific.io-github-autotest-libvirt.vcpu_hotpluggable.negative_test.dup_order1: PASS (11.17 s)
 (7/7) type_specific.io-github-autotest-libvirt.vcpu_hotpluggable.negative_test.dup_order2: PASS (11.97 s)
RESULTS    : PASS 7 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0
JOB HTML   : /root/avocado/job-results/job-2017-08-21T17.19-0ac25ce/html/results.html
TESTS TIME : 176.44 s

File: libvirt/tests/src/cpu/vcpu_hotpluggable.py (hunk 9 above)
Comment (Grace-chhu) [inline libvirt/tests/src/cpu/vcpu_hotpluggable.py]: Modified, thank you!

File: libvirt/tests/cfg/cpu/vcpu_hotpluggable.cfg (hunk 8 above)
Comment (Grace-chhu) [inline libvirt/tests/cfg/cpu/vcpu_hotpluggable.cfg]: Modified, thank  you!

File: libvirt/tests/src/cpu/vcpu_hotpluggable.py (hunk 1 above)
Comment (Grace-chhu) [inline libvirt/tests/src/cpu/vcpu_hotpluggable.py]: Modified, thank you!

File: libvirt/tests/src/cpu/vcpu_hotpluggable.py (hunk 6 above)
Comment (Grace-chhu) [inline libvirt/tests/src/cpu/vcpu_hotpluggable.py]: Modified to test.fail, thank you!

File: libvirt/tests/src/cpu/vcpu_hotpluggable.py (hunk 7 above)
Comment (Grace-chhu) [inline libvirt/tests/src/cpu/vcpu_hotpluggable.py]: Modified, thank you!

File: libvirt/tests/src/cpu/vcpu_hotpluggable.py (hunk 4 above)
Comment (Grace-chhu) [inline libvirt/tests/src/cpu/vcpu_hotpluggable.py]: Modified, use check_if_vm_vcpu_match, thank you!

**Comment by waynesun09** [comment]
@Grace-chhu pls make your branch clean, you have bring in one other commit in this PR

File: libvirt/tests/cfg/cpu/vcpu_hotpluggable.cfg [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    vcpus_max = "8"
+    vcpus_enabled = "0,1,3,5"
+    vcpus_hotpluggable = "1,2,3,4,5,6,7,8"
+    vcpus_order = "{'0':'1', '1':'3', '3':'2', '5':'4'}"
+    config_libvirtd = "no"
+    log_file = "libvirtd.log"
+    set_live_vcpus = ""
+    set_config_vcpus = ""
+    variants:
+        - possitive_test:
```
Comment (lento-sun) [inline libvirt/tests/cfg/cpu/vcpu_hotpluggable.cfg]: should be "positive"

Comment (Grace-chhu) [reply libvirt/tests/cfg/cpu/vcpu_hotpluggable.cfg]: Modified, thank you!

**Pull request merged**